bindings/go/blob/filesystem/blob_from_path.go

```diff
@@ -0,0 +1,342 @@
+	// Determine filesystem root and start dir walk based on PreserveDir
+	baseFSPath := path
+	subPath := "."
+	if opt.PreserveDir {
+		// Root the virtual FS at the parent, and start walking at the preserved base directory name.
+		baseFSPath = filepath.Dir(path)
+		subPath = filepath.Base(path)
+	}
+
+	// Create a virtual filesystem rooted at baseFSPath
+	fileSystem, err := NewFS(baseFSPath, os.O_RDONLY)
+	if err != nil {
+		return nil, fmt.Errorf("error creating virtual filesystem for path %q: %w", baseFSPath, err)
+	}
+
+	// Create TAR stream using pipe
+	pr, pw := io.Pipe()
+
+	go func() {
+		tw := tar.NewWriter(pw)
+
+		// Track any error from TAR creation
+		var tarErr error
+
+		defer func() {
+			// Close TAR writer
+			closeErr := tw.Close()
+			// Close pipe with combined error (if any, otherwise nil -> io.EOF)
+			_ = pw.CloseWithError(errors.Join(tarErr, closeErr))
+		}()
+
+		// If context is already done, abort early.
+		select {
+		case <-ctx.Done():
+			tarErr = fmt.Errorf("context cancelled while preparing tar for path %q: %w", path, ctx.Err())
+			return
+		default:
+		}
+
+		// Create TAR content from directory
+		tarErr = createTarFromDir(ctx, fileSystem, subPath, opt, tw)
+	}()
+
+	// Set media type for TAR blob
+	mediaType := DefaultTarMediaType
+	if opt.MediaType != "" {
+		mediaType = opt.MediaType
+	}
+
+	// Create ReadOnlyBlob from TAR stream
+	var tarBlob blob.ReadOnlyBlob = direct.New(pr, direct.WithMediaType(mediaType))
+
+	// Apply compression if requested
+	if opt.Compress {
+		tarBlob = compression.Compress(tarBlob)
+	}
+	return tarBlob, nil
+}
+
+// createSingleFileBlob creates a blob from the specified single file.
+func createSingleFileBlob(path string, opt DirOptions) (blob.ReadOnlyBlob, error) {
+	// Validate that include/exclude patterns are not used for single files
+	if len(opt.IncludePatterns) > 0 || len(opt.ExcludePatterns) > 0 {
+		return nil, fmt.Errorf("include/exclude patterns are not supported for single files")
+	}
+
+	// Open file for reading
+	fr, err := os.Open(path)
+	if err != nil {
+		return nil, fmt.Errorf("error opening single file %q for reading: %w", path, err)
+	}
+
+	// Determine media type for single file
+	mediaType := DefaultFileMediaType
+	if opt.MediaType != "" {
+		mediaType = opt.MediaType
+	}
+
+	// Create ReadOnlyBlob from file
+	var fileBlob blob.ReadOnlyBlob = direct.New(fr, direct.WithMediaType(mediaType))
+	// Apply compression if requested
+	if opt.Compress {
+		fileBlob = compression.Compress(fileBlob)
+	}
+	return fileBlob, nil
+}
+
+// createTarFromDir creates a TAR archive from the filesystem.
+// Uses the virtual filesystem to read the directory contents.
+// Uses fs.WalkDir to traverse the directory structure in a deterministic order.
+// This is required to ensure reproducible TAR archives.
+func createTarFromDir(ctx context.Context, fileSystem FileSystem, subPath string, opt DirOptions, tw *tar.Writer) error {
```

Comment:
for me this function is too long and too complicated. FWIW this should be separated into smaller funcs imho. were not in C land